core/network/impl/stream_engine.hpp

```diff
@@ -0,0 +1,287 @@
+          std::make_shared<ScaleMessageReadWriter>(std::move(stream));
+      read_writer->write(msg, [self{shared_from_this()}](auto &&res) {
+        if (self && !res)
+          self->logger_->error("Could not send message, reason: {}",
+                               res.error().message());
+      });
+    }
+
+    template <typename T>
+    void send(const PeerInfo &peer,
+              const Protocol &protocol,
+              std::shared_ptr<T> msg) {
+      BOOST_ASSERT(msg);
+      BOOST_ASSERT(!protocol.empty());
+
+      std::shared_lock cs(streams_cs_);
+      forSubscriber(peer, protocol, [&](auto type, auto &stream) {
+        if (stream) {
+          send(stream, *msg);
+          return;
+        }
+
+        BOOST_ASSERT(type == PeerType::kReserved);
+        updateStream(peer, protocol, std::move(msg));
+      });
+    }
+
+    template <typename T>
+    void broadcast(const Protocol &protocol, std::shared_ptr<T> msg) {
+      BOOST_ASSERT(msg);
+      BOOST_ASSERT(!protocol.empty());
+
+      std::shared_lock cs(streams_cs_);
+      forEachPeer([&](const auto &peer, auto type, auto &proto_map) {
+        forProtocol(ProtocolDescriptor{.proto_map = proto_map, .type = type},
+                    protocol,
+                    [&](auto stream) {
+                      BOOST_ASSERT(type == PeerType::kReserved || stream);
+                      if (stream) {
+                        send(std::move(stream), *msg);
+                        return;
+                      }
+                      BOOST_ASSERT(type == PeerType::kReserved);
+                      updateStream(peer, protocol, msg);
+                    });
+      });
+    }
+
+    template <typename F>
+    uint32_t count(F &&filter) const {
+      uint32_t c = 0;
+      auto enumerate = [&](const PeerMap &pm) {
+        for (const auto &i : pm) {
+          if (filter(i.first)) c += i.second.size();
+        }
+      };
+      enumerate(syncing_streams_);
+      enumerate(reserved_streams_);
+      return c;
+    }
+
+   private:
+    Host &host_;
+    common::Logger logger_;
+
+    std::shared_mutex streams_cs_;
+    PeerMap reserved_streams_;
+    PeerMap syncing_streams_;
+
+    PeerInfo from(PeerId &pid) const {
+      return PeerInfo{.id = pid, .addresses = {}};
+    }
+
+    PeerInfo from(PeerId &&pid) const {
+      return PeerInfo{.id = std::move(pid), .addresses = {}};
+    }
+
+    outcome::result<PeerInfo> from(std::shared_ptr<Stream> &stream) const {
+      BOOST_ASSERT(stream);
+      auto peer_id_res = stream->remotePeerId();
+      if (!peer_id_res.has_value()) {
+        logger_->error("Can't get peer_id: {}", peer_id_res.error().message());
+        return peer_id_res.error();
+      }
+      return from(std::move(peer_id_res.value()));
+    }
+
+    void uploadStream(std::shared_ptr<Stream> &dst,
+                      std::shared_ptr<Stream> src) {
+      BOOST_ASSERT(src);
+      if (dst) dst->reset();
+      dst = std::move(src);
```

Comment:
I guess reset is redundant here.
```suggestion
      dst.swap(src);
```
or just
```suggestion
      dst = std::move(src);
```

Reply:
Reset not for ```sptr```, but for ```stream```.

Reply:
Ouch... Okay. Was wrong.